pkg/virt-launcher/virtwrap/live-migration-source.go

```diff
@@ -918,6 +918,10 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 		if err != nil {
 			log.Log.Object(vmi).Reason(err).Error("Live migration failed.")
 			migrationErrorChan <- err
+
+			if err := l.hotPlugHostDevices(vmi); err != nil {
+				log.Log.Error("failed to hot-plug host-devices after migration failure")
```

Comment:
since this `hotPlugHostDevices` returns an error, indicating that it could technically fail for some reason, it probably needs to go somewhere that it can be retried.

why not put this logic in SyncVMI(). That way virt-handler will reconcile it and retry if errors occur. I believe we're guaranteed that SyncVMI will get called after a migration failure due to the vmi status update.

Reply:
@davidvossel Hey thanks for the feedback!
`hotPlugHostDevices` is disruptive operation that could have side effects, thus call it only on specific place where we know we want it to be invoked instead of placing it on SyncVMI directly.

It does make sense for me to move it to SyncVMI as you suggest, I prefer to investigate it first and do it on a follow up PR. Moving the logic to SyncVMI changes the behavior of this feature and I want to make sure we do it right :)

Reply:
We need a solution right out of the gate where this re-hotplug is resilient to failure.

Reply:
@davidvossel When attaching SRIOV VF to the guest fails we can merely raise log/event message about it from kubevirt perspective, I mean we cant fix it on the fly.
Troubleshooting it will probably require a change to the correspond SRIOV PF or on the node
that will probably disrupt the VM.

We didn't want the hotplug on post-migration whether it fails or not to disrupt the migration.
If I got it right, In terms of resilience the VM keep working so I guess it fine, isnt it?

Reply:
> @davidvossel When attaching SRIOV VF to the guest fails we can merely raise log/event message about it from kubevirt perspective, I mean we cant fix it on the fly.

why not? is there no possibility the failure is transient?

It's fairly common to experience unexpected transient errors while under load, which clear up and successfully execute later on.  If errors aren't handled in an eventually consistent fashion, then those errors would have no chance of being automatically reconciled.

Reply:
@davidvossel , I think your suggestion to reconcile the desired state in regards to the actual state of the domain is ideal, but in this specific case may not be trivial.
- The operation of hot-{un}plug is intrusive to the guest (it mutates the guest), which may result in unknown outcomes to application running there if done too often.
- The action itself is limited to a migration, and we have not generalize it. Therefore, as part of an upgrade, devices are hot-unplug and hot-plugged on the target, but with the current implementation, we do not try to recover.
- The suggestion in the PR is to try and recover the devices in case of migration failure, but still keep it limited to the migration scope only. It is a best-effort operation for now.

If we want to generalize it, and in the reconcile loop to make the hot-{un}plug, then we can move and combine this action of recovery and the action on the target node when migration is finalized.

My suggestion is to bring this forward as a change to the current migration-specific scenario and consider the implications of doing it in that scope.
Keeping the current small scoped domain of migration to one-timer attempts to {un}plug for now.